src/Core/Managed/Shared/Extensibility/MetricManager.cs

```diff
@@ -0,0 +1,285 @@
+        private bool disposed;
+
+        /// <summary>
+        /// A dictionary of all metrics instantiated via this manager.
+        /// </summary>
+        private ConcurrentDictionary<Metric, SimpleMetricStatisticsAggregator> metricDictionary;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricManager"/> class.
+        /// </summary>
+        public MetricManager()
+            : this(new TelemetryClient())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricManager"/> class.
+        /// </summary>
+        /// <param name="client">Telemetry client to use to output aggregated metric data.</param>
+        public MetricManager(TelemetryClient client)
+        {
+            this.telemetryClient = client ?? new TelemetryClient();
+            this.metricDictionary = new ConcurrentDictionary<Metric, SimpleMetricStatisticsAggregator>();
+
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            this.snapshotTimer = new TaskTimer() { Delay = GetWaitTime() };
+            this.snapshotTimer.Start(this.SnapshotAndReschedule);
+        }
+
+        /// <summary>
+        /// Gets a list of metric processors associated
+        /// with this instance of <see cref="MetricManager"/>.
+        /// </summary>
+        internal IList<IMetricProcessor> MetricProcessors
+        {
+            get
+            {
+                TelemetryConfiguration config = this.telemetryClient.TelemetryConfiguration;
+
+                return config.MetricProcessors;
+            }
+        }
+
+        /// <summary>
+        /// Creates metric.
+        /// </summary>
+        /// <param name="name">Name of the metric.</param>
+        /// <param name="dimensions">Optional dimensions.</param>
+        /// <returns>Metric instance.</returns>
+        public Metric CreateMetric(string name, IDictionary<string, string> dimensions = null)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            return new Metric(this, name, dimensions);
+        }
+
+        /// <summary>
+        /// Flushes the in-memory aggregation buffers.
+        /// </summary>
+        public void Flush()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            this.Flush(false);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            this.disposed = true;
+
+            if (this.snapshotTimer != null)
+            {
+                this.snapshotTimer.Dispose();
+                this.snapshotTimer = null;
+            }
+
+            this.Flush(true);
+        }
+
+        internal SimpleMetricStatisticsAggregator GetStatisticsAggregator(Metric metric)
```

Comment:
Under what circumstances is the can the manager be disposed. GetStatisticsAggregator is called from the public Track in metric. Should Track ever throw? And if it may, then is this relationship obvious to the user?